Add RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors

youennf opened the pull request:
#### f9eeaaf81af518fb70a641e694c3d99dc150b970
<pre>
Add RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors
<a href="https://rdar.apple.com/149541424">rdar://149541424</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291740">https://bugs.webkit.org/show_bug.cgi?id=291740</a>

Reviewed by Jean-Yves Avenard.

We add support for RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors.
To match the spec, we do not process RTCEncodedAudioFrame and RTCEncodedVideoFrame when enqueued on a writable if they were not created by the corresponding readable.

We use libwebrtc routines to update metadata as part of RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors.
We only partially support metadata changes, based on libwebrtc support and skip other potential changes.

* LayoutTests/imported/w3c/web-platform-tests/interfaces/webrtc-encoded-transform.idl:
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-clone.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-clone.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-constructors.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-constructors.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-worker.js: Added.
(onrtctransform.process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/routines.js:
(async createConnectionWithTransform):
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/RTCEncodedAudioFrame.cpp:
(WebCore::RTCEncodedAudioFrame::create):
* Source/WebCore/Modules/mediastream/RTCEncodedAudioFrame.h:
* Source/WebCore/Modules/mediastream/RTCEncodedAudioFrame.idl:
* Source/WebCore/Modules/mediastream/RTCEncodedVideoFrame.cpp:
(WebCore::RTCEncodedVideoFrame::create):
* Source/WebCore/Modules/mediastream/RTCEncodedVideoFrame.h:
* Source/WebCore/Modules/mediastream/RTCEncodedVideoFrame.idl:
* Source/WebCore/Modules/mediastream/RTCRtpScriptTransformer.cpp:
(WebCore::RTCRtpScriptTransformer::writable):
(WebCore::RTCRtpScriptTransformer::start):
* Source/WebCore/Modules/mediastream/RTCRtpTransformableFrame.h:
(WebCore::RTCRtpTransformableFrame::isFromTransformer const):
(WebCore::RTCRtpTransformableFrame::setTransformer):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
(WebCore::LibWebRTCRtpTransformableFrame::setOptions):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/293951@main">https://commits.webkit.org/293951@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69928362095a1918a01fa5220204bf5ac478d89a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19746 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105228 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50679 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102139 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28219 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76198 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33269 "Found 60 new test failures: animations/3d/animate-to-transform-perpective-to-none.html animations/3d/change-transform-in-end-event.html animations/3d/full-rotation-animation.html animations/3d/replace-filling-transform.html animations/3d/state-at-end-event-transform.html animations/3d/transform-origin-vs-functions.html animations/3d/transform-perspective.html animations/CSSKeyframesRule-name-null.html animations/CSSKeyframesRule-parameters.html animations/added-while-suspended.html ... (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15320 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90415 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56559 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15135 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8412 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50049 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85046 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8495 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107587 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27211 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85153 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27574 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86621 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21598 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29362 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7110 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21062 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27148 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32377 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26959 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30275 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28518 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
```diff
@@ -138,6 +138,47 @@ Ref<RTCRtpTransformableFrame> LibWebRTCRtpTransformableFrame::clone()
+    auto rtcMetadata = static_cast<webrtc::TransformableVideoFrameInterface*>(m_rtcFrame.get())->Metadata();
+
+    if (newMetadata.frameId)
+        rtcMetadata.SetFrameId(*newMetadata.frameId);
```
can't this be done in `VideoFrameMetadata` instead? 
e.g. `rtcMetadata.setWith(newMetadata)`

youennf replied:
rtcMetadata is libwebrtc and if some fields are not set, we are keeping the current rtc field values.

jyavenard reviewed (approved):
r=me with question

youennf committed:
Add RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors
rdar://149541424
https://bugs.webkit.org/show_bug.cgi?id=291740

Reviewed by Jean-Yves Avenard.

We add support for RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors.
To match the spec, we do not process RTCEncodedAudioFrame and RTCEncodedVideoFrame when enqueued on a writable if they were not created by the corresponding readable.

We use libwebrtc routines to update metadata as part of RTCEncodedAudioFrame and RTCEncodedVideoFrame constructors.
We only partially support metadata changes, based on libwebrtc support and skip other potential changes.

* LayoutTests/imported/w3c/web-platform-tests/interfaces/webrtc-encoded-transform.idl:
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-clone.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-clone.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-constructors.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-constructors.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/encoded-frame-worker.js: Added.
(onrtctransform.process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/routines.js:
(async createConnectionWithTransform):
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/RTCEncodedAudioFrame.cpp:
(WebCore::RTCEncodedAudioFrame::create):
* Source/WebCore/Modules/mediastream/RTCEncodedAudioFrame.h:
* Source/WebCore/Modules/mediastream/RTCEncodedAudioFrame.idl:
* Source/WebCore/Modules/mediastream/RTCEncodedVideoFrame.cpp:
(WebCore::RTCEncodedVideoFrame::create):
* Source/WebCore/Modules/mediastream/RTCEncodedVideoFrame.h:
* Source/WebCore/Modules/mediastream/RTCEncodedVideoFrame.idl:
* Source/WebCore/Modules/mediastream/RTCRtpScriptTransformer.cpp:
(WebCore::RTCRtpScriptTransformer::writable):
(WebCore::RTCRtpScriptTransformer::start):
* Source/WebCore/Modules/mediastream/RTCRtpTransformableFrame.h:
(WebCore::RTCRtpTransformableFrame::isFromTransformer const):
(WebCore::RTCRtpTransformableFrame::setTransformer):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
(WebCore::LibWebRTCRtpTransformableFrame::setOptions):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/293951@main
